pkg/gardenlet/bootstrap/certificate/certificate_rotation_test.go

```diff
@@ -0,0 +1,316 @@
+
+		ctrl            *gomock.Controller
+		k8sGardenClient *mock.MockInterface
+		k8sClient       client.Client
+
+		seedName = "test"
+		seed     *gardencorev1beta1.Seed
+
+		gardenClientConnection = &config.GardenClientConnection{
+			KubeconfigSecret: &corev1.SecretReference{
+				Name:      "gardenlet-kubeconfig",
+				Namespace: "garden",
+			},
+		}
+
+		approvedCSR = certificatesv1beta1.CertificateSigningRequest{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "watched-csr",
+			},
+			Status: certificatesv1beta1.CertificateSigningRequestStatus{
+				Conditions: []certificatesv1beta1.CertificateSigningRequestCondition{
+					{
+						Type: certificatesv1beta1.CertificateApproved,
+					},
+				},
+				Certificate: []byte("my-cert"),
+			},
+		}
+
+		deniedCSR = certificatesv1beta1.CertificateSigningRequest{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "watched-csr",
+			},
+			Status: certificatesv1beta1.CertificateSigningRequestStatus{
+				Conditions: []certificatesv1beta1.CertificateSigningRequestCondition{
+					{
+						Type: certificatesv1beta1.CertificateDenied,
+					},
+				},
+			},
+		}
+	)
+
+	BeforeEach(func() {
+		ctrl = gomock.NewController(GinkgoT())
+		k8sGardenClient = mock.NewMockInterface(ctrl)
+
+		seed = &gardencorev1beta1.Seed{
+			ObjectMeta: metav1.ObjectMeta{Name: seedName},
+		}
+
+		// create one fake client to both talk to the Seed API server and Garden API server
+		s := runtime.NewScheme()
+
+		// required schemes
+		Expect(corev1.AddToScheme(s)).ToNot(HaveOccurred())
+		Expect(gardencorescheme.AddToScheme(s)).ToNot(HaveOccurred())
+
+		k8sClient = fakeclient.NewFakeClientWithScheme(s, seed)
+		k8sGardenClient.EXPECT().Client().Return(k8sClient).AnyTimes()
+	})
+
+	AfterEach(func() {
+		ctrl.Finish()
+	})
+
+	Describe("#rotateCertificate", func() {
+		It("should not return an error", func() {
+			kubeClient := fake.NewSimpleClientset(&approvedCSR)
+			k8sGardenClient.EXPECT().Kubernetes().Return(kubeClient)
+
+			// mock the update of the secret in the seed with the rotated kubeconfig
+			certClientConfig := &rest.Config{Host: "testhost", TLSClientConfig: rest.TLSClientConfig{
+				Insecure: false,
+				CAFile:   "filepath",
+			}}
+
+			k8sGardenClient.EXPECT().RESTConfig().Return(certClientConfig)
+			k8sGardenClient.EXPECT().DirectClient().Return(k8sClient)
+
+			fakeClientMap := fakeclientmap.NewClientMap().
+				AddClient(keys.ForGarden(), k8sGardenClient).
+				AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+			err := rotateCertificate(context.TODO(), log, gardenClientConnection, fakeClientMap, seedName)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("should return an error - CSR is denied", func() {
+			kubeClient := fake.NewSimpleClientset(&deniedCSR)
+			k8sGardenClient.EXPECT().Kubernetes().Return(kubeClient)
+
+			fakeClientMap := fakeclientmap.NewClientMap().
+				AddClient(keys.ForGarden(), k8sGardenClient).
+				AddClient(keys.ForSeed(seed), k8sGardenClient)
```

Comment:
Why is `k8sGardenClient` added as a seed client to the client map?

Reply:
One client that can be used for both the Seed as well as the Garden cluster as [noted here](https://github.com/gardener/gardener/blob/67833d9f9bbfd0b9a8ddd0cfffec6a455ead2586/pkg/gardenlet/bootstrap/certificate/certificate_rotation_test.go#L103) for simplicity reasons as in tests we do not really need to separate clients.

Reply:
Well, I think one point of using mocks in unit tests, is to make sure, that the tested functions call the right stuff with the right parameters and react correct on results. 
So I disagree, that we don't need seperate clients, as we want our tests to verify the correct behaviour of the tested function, which would be using the right clients for the right calls.

Reply:
@rfranzke @timuthy Can you state your opinion on this?

Reply:
Makes sense what you say @timebertt. This way the test would still pass if you changed the implementation wrongly.

Reply:
Thanks guys - makes a lot of sense, I adjusted the implementation to use a separate seed client. Also using a mock client instead of fake client to be able to include more mocks.